app/src/app/actions/testConnection.ts

```diff
@@ -0,0 +1,211 @@
+'use server';
+
+import * as z from 'zod';
+
+type TestConnectionResult = {
+  success: boolean;
+  message: string;
+  service: string;
+};
+
+const testConnectionSchema = z.discriminatedUnion('service', [
+  z.object({
+    service: z.literal('claude'),
+    claudeKey: z.string().min(1),
+  }),
+  z.object({
+    service: z.literal('openai'),
+    openaiKey: z.string().min(1),
+  }),
+  z.object({
+    service: z.literal('reddit'),
+    redditClientId: z.string().min(1),
+    redditClientSecret: z.string().min(1),
+  }),
+  z.object({
+    service: z.literal('supabase'),
+    supabaseProjectUrl: z.string().min(1),
+    supabaseApiKey: z.string().min(1),
+  }),
+  z.object({
+    service: z.literal('osf'),
+    osfApiKey: z.string().min(1),
+  }),
+]);
+
+export type TestConnectionService = z.infer<
+  typeof testConnectionSchema
+>['service'];
+
+export async function testConnection(
+  prevState: TestConnectionResult | undefined,
+  formData: FormData,
+): Promise<TestConnectionResult | undefined> {
+  const rawData = Object.fromEntries(formData);
+  const parsed = testConnectionSchema.safeParse(rawData);
+
+  if (!parsed.success) {
+    return {
+      success: false,
+      message: 'Invalid request',
+      service: '',
```

Comment:
Don't like this empty string, how about undefined or null. But actually why do we need to return the service?